### PR TITLE
feat(EMS-697): Your buyer - Buyer company - country field

### PR DESCRIPTION
**Pull request by vensurag**
This PR adds the "Countries select" field to the Your buyer.

**Summary of changes :**

1) Create countries SelectOption as per GOV design structure.
2) Implemented validation for countries select as per requirement.
3) Created field and page variables required for countries select and validations. 
4) Create a mapCountries select helper function to map countries for select.
5) Created routes structure of Your-buyer
6) Add countries fixtures to E2E tests.

### CODE REVIEW

File: src/ui/server/constants/field-ids/insurance/your-buyer/index.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const yourBuyerPageVariables = () => ({
+  FIELDS: {
+    BUYER_COUNTRY: {
+      ID: 'countryId',
+      NAME: 'countryName',
+      LABEL: 'Country',
+      ERROR_MESSAGE: 'Select your buyer country',
+    },
+  },
+});
```
Comment (ttbarnes) [inline src/ui/server/constants/field-ids/insurance/your-buyer/index.ts:10]: Field IDs should be purely field IDs... Page variables should be in the controller, but pulling things in from constants/content strings.

- Labels should come from field content strings in: `content-strings/fields/insurance/your-buyer`
- Error message should come from content strings: `content-strings/error-messages`
- `NAME` shouldn't be required

File: src/ui/server/controllers/insurance/your-buyer/your-buyer-details/index.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+export const get = async (req: Request, res: Response) => {
+  if (!req.session.submittedData || !req.session.submittedData.insuranceEligibility) {
+    req.session.submittedData = {
+      ...req.session.submittedData,
+      insuranceEligibility: {},
+    };
+  }
+
+  const countries = await api.external.getCountries();
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/your-buyer-details/index.ts:18]: Can we use the new API call that gets the countries from keystone API please instead of external?

```js
const countries = await api.keystone.countries.getAll();
```

File: src/ui/server/controllers/insurance/your-buyer/your-buyer-details/index.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { mapCountriesSelect } from '../../../../helpers/mappings/map-countries-select';
+import { yourBuyerDetailsValidation } from './validation/yourBuyerDetailsValidation';
+
+export const get = async (req: Request, res: Response) => {
+  if (!req.session.submittedData || !req.session.submittedData.insuranceEligibility) {
+    req.session.submittedData = {
+      ...req.session.submittedData,
+      insuranceEligibility: {},
+    };
+  }
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/your-buyer-details/index.ts:16]: No need for this, it's is for the eligibility flow only

File: src/ui/server/controllers/insurance/your-buyer/your-buyer-details/validation/yourBuyerDetailsValidation.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+export const yourBuyerDetailsValidation = (responseBody: RequestBody) => {
+  const { countryName } = responseBody;
+  const YOUR_BUYER_PAGE_VARIABLE = yourBuyerPageVariables();
+  let updatedErrors;
+  if (!countryName) {
+    updatedErrors = generateValidationErrors(YOUR_BUYER_PAGE_VARIABLE.FIELDS.BUYER_COUNTRY.ID, YOUR_BUYER_PAGE_VARIABLE.FIELDS.BUYER_COUNTRY.ERROR_MESSAGE);
+  }
+  return updatedErrors;
+};
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/your-buyer-details/validation/yourBuyerDetailsValidation.ts:13]: Can we use the same pattern/approach as other validation please e.g https://github.com/UK-Export-Finance/exip/tree/main-application/src/ui/server/controllers/insurance/policy-and-export/single-contract-policy/validation

Need a rules directory and just importing field IDs and error messages from constants

File: src/ui/server/index.ts [hunk 5]
```diff
@@ -28,7 +28,7 @@ import getApplication from './middleware/insurance/get-application';
 
 // @ts-ignore
 const app = express();
-const PORT = process.env.PORT || 5000;
+const PORT = process.env.PORT || 4000;
```
Comment (ttbarnes) [inline src/ui/server/index.ts:31]: Can we revert this change

File: src/ui/server/routes/insurance/index.ts [hunk 6]
```diff
@@ -42,5 +43,6 @@ insuranceRouter.get(INSURANCE_ROUTES.PAGE_NOT_FOUND, pageNotFoundGet);
 
 insuranceRouter.use('/', insuranceEligibilityRoutes);
 insuranceRouter.use(`${INSURANCE_ROOT}/:referenceNumber`, insuranceBusinessRouter);
+insuranceRouter.use(`${INSURANCE_ROOT}/`, yourBuyerRouter);
```
Comment (ttbarnes) [inline src/ui/server/routes/insurance/index.ts:46]: We need to have the dynamic parameter here `:referenceNumber`

File: src/ui/server/routes/insurance/index.ts [hunk 7]
```diff
@@ -11,6 +11,7 @@ import { post as typeOfPolicySaveAndBackPost } from '../../controllers/insurance
 import { get as pageNotFoundGet } from '../../controllers/insurance/page-not-found';
 import insuranceEligibilityRoutes from './eligibility';
 import insuranceBusinessRouter from './business';
+import yourBuyerRouter from './your-buyer-details';
```
Comment (ttbarnes) [inline src/ui/server/routes/insurance/index.ts:14]: The file should be called the same, `your-buyer`

Buyer details is a page/route within the "your buyer" section

File: src/ui/server/routes/insurance/your-buyer-details/index.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import express from 'express';
+import { YOUR_BUYER } from '../../../constants/routes/insurance/your-buyer';
+import { get as getYourBuyerOrganisation, post as postYourBuyerOrganisation } from '../../../controllers/insurance/your-buyer/your-buyer-details';
```
Comment (ttbarnes) [inline src/ui/server/routes/insurance/your-buyer-details/index.ts:3]: Shoudl it be called `getYourBuyerDetails` instead of `getYourBuyerOrganisation`?

File: src/ui/templates/insurance/your-buyer/your-buyer-details.njk [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+  <form method="POST" novalidate="novalidate">
+
+    <div class="govuk-grid-row">
+      <div class="govuk-grid-column-one-half-from-desktop">
+
+        {% set class = "govuk-form-group" %}
+
+        {% if validationErrors.errorList[FIELD_ID] %}
+          {% set class = "govuk-form-group govuk-form-group--error" %}
+        {% endif %}
```
Comment (ttbarnes) [inline src/ui/templates/insurance/your-buyer/your-buyer-details.njk:40]: No need for the class declerations

File: src/ui/templates/insurance/your-buyer/your-buyer-details.njk [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+        {{ govukSelect({
+                id: FIELDS.BUYER_COUNTRY.ID,
+                name: FIELDS.BUYER_COUNTRY.NAME,
+                label: {
+                  text: FIELDS.BUYER_COUNTRY.LABEL
+                },
+                items: countries
+              }) }}
```
Comment (ttbarnes) [inline src/ui/templates/insurance/your-buyer/your-buyer-details.njk:49]: indentation

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    heading().invoke('text').then((text) => {
+      expect(text.trim()).equal(CONTENT_STRINGS.PAGE_TITLE);
+    });
+  });
+
+  it('renders an intro paragraph', () => {
+    insurance.policyAndExport.typeOfPolicyPage.intro().invoke('text').then((text) => {
+      expect(text.trim()).equal(CONTENT_STRINGS.INTRO);
+    });
+  });
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js:122]: This is testing the type of policy page - needs to be only the your buyer page i.e a test that the new form field and label is rendered.

Will need to create a new yourBuyerPage in the E2E pages definition.

Then can remove the page intro and intro paragraph test

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    cy.visit(ROUTES.INSURANCE.START, {
+      auth: {
+        username: Cypress.config('basicAuthKey'),
+        password: Cypress.config('basicAuthSecret'),
+      },
+    });
+
+    cy.submitInsuranceEligibilityAndStartApplication();
+
+    taskList.prepareApplication.tasks.policyTypeAndExports.link().click();
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js:50]: We can remove this (needs to be updated in a future PR to click into "your buyer" section).

Instead for now, we can remove this and use the `goToPageDirectly` function to get to the your buyer

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+const goToPageDirectly = (referenceNumber) => {
+  cy.visit(`${INSURANCE_ROOT}/${referenceNumber}${ROUTES.INSURANCE.YOUR_BUYER.YOUR_BUYER_DETAILS}`, {
+    auth: {
+      username: Cypress.config('basicAuthKey'),
+      password: Cypress.config('basicAuthSecret'),
+    },
+  });
+};
+
+context('Insurance - Your Buyer - Type of yourbuyer Page - As an exporter, I want to enter the type of Your Buyer details I need for my export contract', () => {
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js:37]: typo

File: src/ui/server/controllers/insurance/your-buyer/index.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import { ROUTES, TEMPLATES } from '../../../constants';
+import insuranceCorePageVariables from '../../../helpers/page-variables/core/insurance';
+import { mapCountriesSelect } from '../../../helpers/mappings/map-countries-select';
+import yourBuyerDetailsValidation from './validation/yourBuyerDetailsValidation';
+import { yourBuyerFiledVariables } from '../../../content-strings/fields/insurance/your-buyer';
+
+const getMappedCountries = async (res: Response) => {
+  const countries = await api.external.getCountries();
+  if (!countries || !countries.length) {
+    return res.redirect(ROUTES.PROBLEM_WITH_SERVICE);
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/index.ts:13]: Do you think we should have the `res.redirect` logic in the get/post itself? So we're not passing `res` around? Could return false in this function to flag if we need to redirect in the controller

File: src/ui/server/controllers/insurance/your-buyer/index.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    }),
+    ...yourBuyerFiledVariables,
+    countries: mappedCountries,
+  });
+};
+
+export const post = async (req: Request, res: Response) => {
+  const mappedCountries = await getMappedCountries(res);
+  const validationErrors = yourBuyerDetailsValidation(req.body);
+  if (validationErrors && Object.keys(validationErrors).length) {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/index.ts:33]: Very minor can we add a space before `if (validationErrors ...` so it's easier to read, not a blocker

File: src/ui/server/controllers/insurance/your-buyer/validation/yourBuyerDetailsValidation.test.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import yourBuyerDetailsValidation from './yourBuyerDetailsValidation';
+import validationRules from './rules/rules';
+import combineValidationRules from '../../../../helpers/combine-validation-rules';
+
+describe('controllers/insurance/policy-and-export/single-contract-policy/validation', () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/validation/yourBuyerDetailsValidation.test.ts:5]: path doesn't match what file we're in

File: src/ui/server/controllers/insurance/your-buyer/validation/yourBuyerDetailsValidation.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import { RequestBody, ValidationErrors } from '../../../../../types';
+import combineValidationRules from '../../../../helpers/combine-validation-rules';
+import validationRules from './rules/rules';
+
+const yourBuyerDetailsValidation = (formBody: RequestBody): ValidationErrors => combineValidationRules(validationRules, formBody);
+
+export default yourBuyerDetailsValidation;
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/validation/yourBuyerDetailsValidation.ts:7]: Can this file be called `index.ts` please like other validation functions

File: src/ui/server/routes/insurance/index.test.ts [hunk 18]
```diff
@@ -21,8 +21,8 @@ describe('routes/insurance', () => {
   });
 
   it('should setup all routes', () => {
-    expect(get).toHaveBeenCalledTimes(25);
-    expect(post).toHaveBeenCalledTimes(22);
+    expect(get).toHaveBeenCalledTimes(26);
+    expect(post).toHaveBeenCalledTimes(23);
```
Comment (ttbarnes) [inline src/ui/server/routes/insurance/index.test.ts:25]: Can we add the get and post tests for the new routes please? e.g

```js
expect(get).toHaveBeenCalledWith( ....
expect(post).toHaveBeenCalledWith( ....
```

File: src/ui/server/test-mocks/mock-ciscountries.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+  {
+    isoCode: 'B',
+    marketName: 'Mock B',
+    shortTermCoverAvailabilityDesc: API.CIS.SHORT_TERM_COVER_AVAILABLE.YES,
+    ESRAClasificationDesc: 'nbiIssueAvailable',
+    NBIIssue: 'NBIIssue',
+  },
+] as Array<CisCountry>;
+
+export default mockCisCountries;
```
Comment (ttbarnes) [inline src/ui/server/test-mocks/mock-ciscountries.ts:21]: We already have a mock file for CIS countries? `mock-cis-countries`

File: src/ui/templates/insurance/your-buyer/your-buyer-details.njk [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+      errorList: validationErrors.summary
+    }) }}
+  {% endif %}
+
+  <form method="POST" novalidate="novalidate">
+    <input type="hidden" name="_csrf" value="{{ csrfToken }}">
+    <div class="govuk-grid-row">
+      <div class="govuk-grid-column-one-half-from-desktop">
+
+        {% set class = "govuk-form-group" %}
```
Comment (ttbarnes) [inline src/ui/templates/insurance/your-buyer/your-buyer-details.njk:36]: This can be removed as it's not used

File: src/ui/server/helpers/mappings/map-countries-select.ts [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+      value: '',
+    };
+
+    const result = [defaultOption, ...sorted];
+
+    return result;
+  }
+
+  return sorted;
+};
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-countries-select.ts:42]: Test coverage - can probably grab most of the required tests from `map-currencies` and shuffle some things as they're both very similar

File: src/ui/server/helpers/mappings/map-countries-select.test.ts [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    {
+      marketName: 'Algeria',
+      isoCode: 'DZA',
+      oecdRiskCategory: 2,
+      shortTermCoverAvailabilityDesc: 'D',
+      ESRAClasificationDesc: 'E',
+      NBIIssue: 'F',
+    },
+  ] as Array<CisCountry>;
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-countries-select.test.ts:21]: Can we use countries from the test mocks?

File: src/ui/server/helpers/mappings/map-countries-select.test.ts [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+
+  describe('getSupportedCurrencies', () => {
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-countries-select.test.ts:23]: There's no function called `getSupportedCurrencies` in this file

Comment (ttbarnes) [reply src/ui/server/helpers/mappings/map-countries-select.test.ts:23]: Also this doesn't reflect what the code is actually doing `it('should only return supported currencies (GBP, EUR, USD)', () => {`

File: src/ui/server/helpers/mappings/map-countries-select.test.ts [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+          text: 'Abu Dhabi',
+          value: 'XAD',
+        },
+        {
+          text: 'Algeria',
+          value: 'DZA',
+        },
+      ];
+
+      expect(result[0]).toEqual(expected[0]);
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-countries-select.test.ts:43]: Can we test the whole array rather than one object

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+
+    headingCaption().invoke('text').then((text) => {
+      expect(text.trim()).equal(CONTENT_STRINGS.HEADING_CAPTION);
+    });
+
+    heading().invoke('text').then((text) => {
+      expect(text.trim()).equal(CONTENT_STRINGS.PAGE_TITLE);
+    });
+  });
+
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js:108]: Looks there isn't a page title or heading in the page yet?

Can we add a test for the countries field and label is rendering please? Like in here https://github.com/UK-Export-Finance/exip/blob/main-application/e2e-tests/cypress/e2e/journeys/insurance/policy-and-exports/single-contract-policy/single-contract-policy.spec.js#L220 but we don't need to mention "supported" countries as all countries will be rendered

File: e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+  it('renders a back link with correct url', () => {
+    partials.backLink().should('exist');
+    partials.backLink().invoke('text').then((text) => {
+      expect(text.trim()).equal(LINKS.BACK);
+    });
+
+    partials.backLink().click();
+
+    const expectedUrl = `${Cypress.config('baseUrl')}${INSURANCE_ROOT}/${referenceNumber}${ROUTES.INSURANCE.YOUR_BUYER.YOUR_BUYER_DETAILS}`;
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/your-buyer/your-buyer.spec.js:77]: Test is failing because `YOUR_BUYER.YOUR_BUYER_DETAILS` is undefined

File: src/ui/server/content-strings/error-messages.ts [hunk 27]
```diff
@@ -17,6 +17,9 @@ export const ERROR_MESSAGES = {
   [FIELD_IDS.CURRENCY]: {
     IS_EMPTY: 'Select currency',
   },
+  [FIELD_IDS.COUNTRY]: {
+    IS_EMPTY: 'Select country',
```
Comment (Zainzzkk) [inline src/ui/server/content-strings/error-messages.ts:21]: does this error message align with the figma?

File: src/ui/server/controllers/insurance/your-buyer/index.test.ts [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+import { get, post } from '.';
+import { mapCountriesSelect } from '../../../helpers/mappings/map-countries-select';
+import { PAGES } from '../../../content-strings';
+import { FIELDS } from '../../../content-strings/fields/insurance/your-buyer';
+import insuranceCorePageVariables from '../../../helpers/page-variables/core/insurance';
+
+describe('controllers/insurance/your-buyer/your-buyer-details', () => {
+  let req: Request;
+  let res: Response;
+  //   let refNumber: number;
```
Comment (Zainzzkk) [inline src/ui/server/controllers/insurance/your-buyer/index.test.ts:14]: please remove comment if not using this

File: src/ui/server/controllers/insurance/your-buyer/index.test.ts [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+  let getCountriesSpy = jest.fn(() => Promise.resolve(mockCisCountries));
+
+  beforeEach(() => {
+    req = mockReq();
+    res = mockRes();
+
+    res.locals.application = mockApplication;
+    req.params.referenceNumber = String(mockApplication.referenceNumber);
+    // refNumber = Number(mockApplication.referenceNumber);
```
Comment (Zainzzkk) [inline src/ui/server/controllers/insurance/your-buyer/index.test.ts:23]: same as above

File: src/ui/server/controllers/insurance/your-buyer/index.test.ts [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    describe.skip('when there are no validation errors', () => {
+      beforeEach(() => {
+        req.body = validBody;
+      });
+
+      it('should redirect to needs_to_redirect_at_do_you_need_broker', async () => {
+        await post(req, res);
+
+        // const expected = `${INSURANCE_ROOT}/${req.params.referenceNumber}${ABOUT_GOODS_OR_SERVICES}`;
+        // const expected = `${INSURANCE_ROOT}/${req.params.referenceNumber}/${YOUR_BUYER}`;
```
Comment (Zainzzkk) [inline src/ui/server/controllers/insurance/your-buyer/index.test.ts:109]: please remove these

File: src/ui/templates/insurance/your-buyer/your-buyer-details.njk [hunk 31]
```diff
@@ -0,0 +1,70 @@
...
+          {{ govukButton({
+            text: CONTENT_STRINGS.BUTTONS.SAVE_AND_BACK,
+            classes: "govuk-button--secondary",
+            attributes: {
+              'data-cy': 'save-and-back-button'
+            }
+          }) }}
+        </div>
+      </div>
+
```
Comment (Zainzzkk) [inline src/ui/templates/insurance/your-buyer/your-buyer-details.njk:65]: can we remove these empty spaces please and on line 67

**Review by Zainzzkk — COMMENTED** [review]
We also need a description for the PR

**Comment by vensurag** [comment]
Summary of changes
Create mapSelectOption function for mapping an option into the required GOV design structure.
Created error strcuture fields as per design 
Create a mapCountrie sselect  function to map countries, regardless of where the countries are obtained from (our API, external API).

File: e2e-tests/cypress/e2e/pages/insurance/your-buyer.js [hunk 32]
```diff
@@ -0,0 +1,20 @@
+import { FIELDS } from '../../../../content-strings/fields/insurance/your-buyer';
+
+const {
+  YOUR_BUYER: {
+    BUYER_COUNTRY: {
+      ID,
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/pages/insurance/your-buyer.js:6]: Not a blocker for this PR but in the next PR, will need to use `FIELD_iDS` instead of `FIELDS` because they'll be multiple field ids that need to be consumed

File: src/ui/server/controllers/insurance/your-buyer/index.test.ts [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+    beforeEach(() => {
+      getCountriesSpy = jest.fn(() => Promise.resolve(mockCisCountries));
+      api.external.getCountries = getCountriesSpy;
+    });
+
+    const validBody = {
+      CountryId: mockCisCountries[0].isoCode,
+    };
+
+    describe.skip('when there are no validation errors', () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/index.test.ts:98]: Remove `.skip`

**Review by ttbarnes — APPROVED** [review]
